File: backend/app/models/webrtc.py
from enum import Enum
from typing import Optional

from pydantic import BaseModel


class MessageType(str, Enum):
    OFFER = "offer"
    ANSWER = "answer"
    ICE_CANDIDATE = "ice-candidate"
    WELCOME = "welcome"
    ERROR = "error"


class SDPMessage(BaseModel):
    type: MessageType
    sdp: str
    sdpType: str  # "offer" or "answer"


class ICECandidatePayload(BaseModel):
    candidate: str
    sdpMid: Optional[str]
    sdpMLineIndex: Optional[int]


class ICECandidateMessage(BaseModel):
    type: MessageType
    candidate: ICECandidatePayload


class WelcomeMessage(BaseModel):
    type: MessageType
    client_id: str
    timestamp: str  # ISO 8601 UTC string


class ErrorMessage(BaseModel):
    type: MessageType
    message: str


SignalingMessage = SDPMessage | ICECandidateMessage | WelcomeMessage | ErrorMessage

File: mobile/types/webrtc.ts
export enum MessageType {
  OFFER = 'offer',
  ANSWER = 'answer',
  ICE_CANDIDATE = 'ice-candidate',
  WELCOME = 'welcome',
  ERROR = 'error',
}

export interface SDPMessage {
  type: MessageType.OFFER | MessageType.ANSWER;
  sdp: string;
  sdpType: 'offer' | 'answer';
}

export interface ICECandidatePayload {
  candidate: string;
  sdpMid?: string | null;
  sdpMLineIndex?: number | null;
}

export interface ICECandidateMessage {
  type: MessageType.ICE_CANDIDATE;
  candidate: ICECandidatePayload;
}

export interface WelcomeMessage {
  type: MessageType.WELCOME;
  client_id: string;
  timestamp: string;
}

export interface ErrorMessage {
  type: MessageType.ERROR;
  message: string;
}

export type SignalingMessage = SDPMessage | ICECandidateMessage | WelcomeMessage | ErrorMessage;

export type TransportStatus = 'connecting' | 'open' | 'closing' | 'closed';

export interface SignalingTransport {
  status: TransportStatus;
  connect: () => Promise<void>;
  disconnect: () => void;
  send: (msg: SignalingMessage) => void;
  onMessage: (handler: (msg: SignalingMessage) => void) => void;
}
